File: roadmapplugin/0.12/roadmapplugin/roadmap.py
from trac.core import *
from genshi.filters import Transformer
from genshi.builder import tag
from genshi import HTML
from trac.web.api import IRequestFilter, ITemplateStreamFilter
from operator import attrgetter
from trac.util.translation import domain_functions
import pkg_resources

#from trac.ticket.model import *

_, tag_, N_, add_domain = domain_functions('roadmapplugin',  '_', 'tag_', 'N_', 'add_domain')


# copied from RoadmapFilterPlugin.py, see https://trac-hacks.org/wiki/RoadmapFilterPlugin
def get_session_key(name):
        return "roadmap.filter.%s" % name

class SortRoadMap(Component):
    """Shows another checkbox in roadmap view, 
which allows you to sort milestones in descending order of due date."""
    implements(IRequestFilter, ITemplateStreamFilter)
    
    def __init__(self):
        locale_dir = pkg_resources.resource_filename(__name__, 'locale') 
        add_domain(self.env.path, locale_dir)
        
    def pre_process_request(self, req, handler):
        """ overridden from IRequestFilter"""
        return handler
        
    def post_process_request(self, req, template, data, content_type):
        """ overridden from IRequestFilter"""
        if template == 'roadmap.html':
            sort_desc = self._get_sorting(req)
            
            if sort_desc == 'on':
                milestones = data['milestones']
                ms_with_due = []
                ms_wo_due = []
                for m in milestones:
                    if m.due:
                        ms_with_due.append(m)
                    else:
                        ms_wo_due.append(m)
                stats = data['milestone_stats']
                new_stats = []
                new_milestones = sorted(ms_with_due, key=attrgetter('due'), reverse=True)
                new_milestones.extend(ms_wo_due)
                
                for i, m in enumerate(new_milestones):
                    for j, om in enumerate(milestones):
                        if m.name == om.name:
                           new_stats.append(stats[j])
                           continue
                data['milestones'] = new_milestones
                data['milestone_stats'] = new_stats
                #milestones.sort(key=due) 
        return template, data, content_type

    def filter_stream(self, req, method, filename, stream, data):
        if filename == 'roadmap.html':
            sort_desc = self._get_sorting(req)
            html_str = '<div><input type="checkbox" name="sortdesc"'
            if sort_desc == 'on':
                html_str += ' checked'
            html_str += '/> ' + _('Sort descending') + '</div>'
            html = HTML(html_str)
            filter = Transformer('//form[@id="prefs"]/div[@class="buttons"]')
            return stream | filter.before(html)
        return stream
    
    
    def _get_sorting(self, req):
        sessionKey = get_session_key('sortDesc')
        sort_desc= ''
        if req.args.has_key('sortdesc'):
            sort_desc= req.args['sortdesc']
            # make value persistent...
            req.session[sessionKey]= sort_desc
        elif req.session.has_key(sessionKey):
            # use persistent value...
            sort_descresult= req.session[sessionKey]
        return sort_desc
    
# copied from RoadmapFilterPlugin.py, see https://trac-hacks.org/wiki/RoadmapFilterPlugin
class FilterRoadmap(Component):
    """Filters roadmap milestones.

Mainly copied from [https://trac-hacks.org/wiki/RoadmapFilterPlugin RoadmapFilterPlugin]
and modified a bit.

Thanks to daveappendix """
    implements(IRequestFilter, ITemplateStreamFilter)

    def _session(self, req, name, default):
        return req.session.get(get_session_key(name), default)

    def _getFilter(self, req, name):
        sessionKey = get_session_key(name)
        result= ''
        if req.args.has_key(name):
            result= req.args[name]
            # make value persistent...
            req.session[sessionKey]= result
        elif req.session.has_key(sessionKey):
            # use persistent value...
            result= req.session[sessionKey]
        return result

    def _getCheckbox(self, req, name, default):
        sessionKey= get_session_key(name)
        result= False

        if req.args.has_key('user_modification'):
            # User has hit the update button on the form,
            # so update the session data.
            if req.args.has_key(name):
                result= True
            if result:
                req.session[sessionKey]= 'true'
            else:
                req.session[sessionKey]= 'false'
        elif req.session.get(sessionKey, default) == 'true':
            result= True

        return result

    def _matchFilter(self, name, filters):
        # Existing Trac convention says that the following prefixes
        # on the filter do different things:
        #   ~  - contains
        #   ^  - starts with
        #   $  = ends with
        for filter in filters:
            if filter.startswith('^'):
                if name.startswith(filter[1:]):
                    return True
            elif filter.startswith('$'):
                if name.endswith(filter[1:]):
                    return True
            elif filter.startswith('~'):
                if name.find(filter[1:]) >= 0:
                    return True
            elif name == filter:
                return True
        return False

    # IRequestFilter methods

    def pre_process_request(self, req, handler):
        return handler

    def post_process_request(self, req, template, data, content_type):
        if template == 'roadmap.html':
            inc_milestones= self._getFilter(req, 'inc_milestones')
            exc_milestones= self._getFilter(req, 'exc_milestones')
            show_descriptions= self._getCheckbox(req, 'show_descriptions', 'true')

            if inc_milestones != '':
                inc_milestones = [m.strip() for m in inc_milestones.split('|')]
                filteredMilestones = []
                filteredStats= []
                for i in range(len(data['milestones'])):
                    m= data['milestones'][i]
                    if self._matchFilter(m.name, inc_milestones):
                        filteredMilestones.append(m)
                        filteredStats.append(data['milestone_stats'][i])
                data['milestones'] = filteredMilestones
                data['milestone_stats'] = filteredStats
                
            if exc_milestones != '':
                exc_milestones = [m.strip() for m in exc_milestones.split('|')]
                filteredMilestones = []
                filteredStats= []
                for i in range(len(data['milestones'])):
                    m= data['milestones'][i]
                    if not self._matchFilter(m.name, exc_milestones):
                        filteredMilestones.append(m)
                        filteredStats.append(data['milestone_stats'][i])
                data['milestones'] = filteredMilestones
                data['milestone_stats'] = filteredStats

            if not show_descriptions:
                for m in data['milestones']:
                    m.description= ''

        return (template, data, content_type)

    # ITemplateStreamFilter methods

    def filter_stream(self, req, method, filename, stream, data):
        if filename == 'roadmap.html':
            # Insert the new field for entering user names
            filter = Transformer('//form[@id="prefs"]/div[@class="buttons"]')
            return stream | filter.before(self._user_field_input(req))
        return stream

    def _filterBox(self, req, label, name):
        return tag.label(label,
                         tag.input(type= "text",
                                   name= name,
                                   value= self._session(req, name, ''),
                                   style_= "width:60%",
                                   title = _('available prefixes: contains: ~, starts with: ^, ends with: $') ))

    def _toggleBox(self, req, label, name, default):
        if self._session(req, name, default) == 'true':
            checkbox= tag.input(type= 'checkbox',
                                name= name,
                                value='true',
                                checked='checked')
        else:
            checkbox= tag.input(type= 'checkbox',
                                name= name,
                                value= 'true')
        return checkbox + ' ' + tag.label(label, for_= name)

    def _hackedHiddenField(self):
        # Hack: shove a hidden field in so we can tell if the update
        # button has been hit.
        return tag.input(type= 'hidden',
                         name= 'user_modification',
                         value='true')

    def _user_field_input(self, req):
        return tag.div(self._hackedHiddenField()
                       + self._toggleBox(req,
                                         _('Show milestone descriptions'),
                                         'show_descriptions',
                                         'true')
                       + tag.br()
                       + self._filterBox(req, _('Filter: '), "inc_milestones")
#                       + tag.br()
#                       + self._filterBox(req, "Exclude: ", "exc_milestones")
                       )                    

File: traditionalchinesetranslation/0.10/trunk/0.10.x/templates/roadmap.cs
﻿<?cs include "header.cs"?>
<?cs include "macros.cs"?>

<div id="ctxtnav" class="nav"></div>

<div id="content" class="roadmap">
 <h1>開發藍圖</h1>

 <form id="prefs" method="get" action="">
  <div>
   <input type="checkbox" id="showall" name="show" value="all"<?cs
    if:roadmap.showall ?> checked="checked"<?cs /if ?> />
   <label for="showall">顯示已經完成的里程碑</label>
  </div>
  <div class="buttons">
   <input type="submit" value="更新" />
  </div>
 </form>

 <ul class="milestones"><?cs each:milestone = roadmap.milestones ?>
  <li class="milestone">
   <div class="info">
    <h2><a href="<?cs var:milestone.href ?>">里程碑: <em><?cs
      var:milestone.name ?></em></a></h2>
    <p class="date"<?cs
     if:milestone.completed_date ?> title="<?cs var:milestone.completed_date ?>">
      Completed <?cs var:milestone.completed_delta ?> ago<?cs
     elif:milestone.due_date ?> title="<?cs var:milestone.due_date ?>"><?cs
      if:milestone.late ?>
       <strong><?cs var:milestone.due_delta ?> late</strong><?cs
      else ?>
       Due in <?cs var:milestone.due_delta ?><?cs
      /if ?> (<?cs var:milestone.due_date ?>)<?cs
     else ?>>
      日期待定<?cs
     /if ?>
    </p><?cs
    with:stats = milestone.stats ?><?cs
     if:#stats.total_tickets > #0 ?>
      <table class="progress">
       <tr>
        <td class="closed" style="width: <?cs
          var:#stats.percent_closed ?>%"><a href="<?cs
          var:milestone.queries.closed_tickets ?>" title="<?cs
          var:#stats.closed_tickets ?> of <?cs
          var:#stats.total_tickets ?> ticket<?cs
          if:#stats.total_tickets != #1 ?>s<?cs /if ?> closed"></a></td>
        <td class="open" style="width: <?cs
          var:#stats.percent_active ?>%"><a href="<?cs
          var:milestone.queries.active_tickets ?>" title="<?cs
          var:#stats.active_tickets ?> of <?cs
          var:#stats.total_tickets ?> ticket<?cs
          if:#stats.total_tickets != #1 ?>s<?cs /if ?> active"></a></td>
       </tr>
      </table>
      <p class="percent"><?cs var:#stats.percent_closed ?>%</p>
      <dl>
       <dt>已完成項目:</dt>
       <dd><a href="<?cs var:milestone.queries.closed_tickets ?>"><?cs
         var:stats.closed_tickets ?></a></dd>
       <dt>處理中項目:</dt>
       <dd><a href="<?cs var:milestone.queries.active_tickets ?>"><?cs
         var:stats.active_tickets ?></a></dd>
      </dl><?cs
     /if ?><?cs
    /with ?>
   </div>
   <div class="description"><?cs var:milestone.description ?></div>
  </li><?cs
 /each ?></ul><?cs
 if:trac.acl.MILESTONE_CREATE ?>
  <div class="buttons">
   <form method="get" action="<?cs var:trac.href.milestone ?>"><div>
    <input type="hidden" name="action" value="new" />
    <input type="submit" value="Add new milestone" />
   </div></form>
  </div><?cs
 /if ?>

 <div id="help">
  <strong>提示:</strong> 查看 <a href="<?cs
    var:trac.href.wiki ?>/ZhTracRoadmap">里程碑</a> 對本頁面的幫助.
 </div>

</div>
<?cs include:"footer.cs"?>
